jsapp/js/actions/datashare.es6

```diff
@@ -0,0 +1,110 @@
+  detachSource: {children: ['completed', 'failed']},
+  patchSource: {children: ['completed', 'failed']},
+  getAttachedSources: {children: ['completed', 'failed']},
+  getSharingEnabledAssets: {children: ['completed', 'failed']},
+  toggleDataSharing: {children: ['completed', 'failed']},
+  updateColumnFilters: {children: ['completed', 'failed']},
+});
+
+dataShareActions.attachToSource.listen((assetUid, data) => {
+  dataInterface.attachToSource(assetUid, data)
+    .done(dataShareActions.attachToSource.completed)
+    .fail(dataShareActions.attachToSource.failed);
+});
+dataShareActions.attachToSource.failed.listen((response) => {
+  alertify.error(response?.responseJSON || t('Failed to attach to source'));
+});
+
+dataShareActions.detachSource.listen((attachmentUrl) => {
+  dataInterface.detachSource(attachmentUrl)
+    .done(dataShareActions.detachSource.completed)
+    .fail(dataShareActions.detachSource.failed);
+});
+dataShareActions.detachSource.failed.listen((response) => {
+  alertify.error(response?.responseJSON || t('Failed to detach from source'));
+});
+
+dataShareActions.patchSource.listen((attachmentUrl, data) => {
+  dataInterface.patchSource(attachmentUrl, data)
+    .done(dataShareActions.patchSource.completed)
+    .fail(dataShareActions.patchSource.failed)
+});
+dataShareActions.patchSource.failed.listen((response) => {
+  alertify(response?.responseJSON || t('Failed to patch source'));
+});
+
+dataShareActions.getAttachedSources.listen((assetUid) => {
+  dataInterface.getAttachedSources(assetUid)
+    .done((response) => {
+      let allSources = [];
+      response.results.forEach((source) => {
+        let sourceUid = getAssetUIDFromUrl(source.source);
+        allSources.push({
+          sourceName: truncateString(
+            source.source__name,
+            MAX_DISPLAYED_STRING_LENGTH.connect_projects,
+          ),
+          // Source's asset url
+          sourceUrl: source.source,
+          sourceUid: sourceUid,
+          // Fields that the connecting project has selected to import
+          linkedFields: source.fields,
+          filename: truncateFile(
+            source.filename,
+            MAX_DISPLAYED_STRING_LENGTH.connect_projects,
+          ),
+          // Source project attachment endpoint
+          attachmentUrl: source.url,
+        });
+      });
+      dataShareActions.getAttachedSources.completed(allSources);
+    })
+    .fail(dataShareActions.getAttachedSources.failed);
+});
+
+dataShareActions.getSharingEnabledAssets.listen(() => {
+  dataInterface.getSharingEnabledAssets()
+    .done(dataShareActions.getSharingEnabledAssets.completed)
+    .fail(dataShareActions.getSharingEnabledAssets.failed);
+});
+dataShareActions.getSharingEnabledAssets.failed.listen(() => {
+  alertify.error(t('Failed to retrieve sharing enabled assets'));
+});
+
+// The next two actions have the same endpoint but must be handled very
+// differently so we leave them as seperate actions
```

Comment:
> So if I use `actions.dataShare.toggleDataSharing` and the asset changes because of that, wouldn't `actions.dataShare.updateColumnFilters.completed.listen()` be interested in the response as well?

So regardless if the swithch has just been turned on or off, we want an empty `paried-data` array, and the columns are reset to empty (matching backend functionality) after this request goes through with the listener in the modal, so the column filters are also updated.

I have the `updateColumnFilters` listener specifically listen for changes to the checkboxes, and `toggleDataSharing` handles cases where we want to set clear the columns

Reply:
> Maybe this function needs different parameters to ensure it is clearly different from the other one and wouldn't be misused. So let's make them `uid, isShared` and the function would build the `data` object by itself:
> 
> ```js
> const data = JSON.stringify({
>   data_sharing: {
>     enabled: isShared
>     fields: [],
>   }
> });
> ```

Will make a comment for this and add it to the improvements issue :D